File: custom_data_structures/queue.rb
require_relative 'node'

class Queue
  attr_accessor :first, :last

  def initialize
    @first  = nil
    @last   = nil
  end

  def enqueue value
    new_node = Node.new(value)
    if self.any?
      @last.next = new_node
      @last = new_node
    else
      @first = new_node
      @last = new_node
    end
  end

  def dequeue
    return nil if self.empty?
    dequeued_item = first.value
    if @first == @last
      @first = nil
      @last = nil
    else
      @first = first.next
    end
    dequeued_item
  end

  def empty?
    !@first
  end

  def any?
    !!@first
  end
end

# a = Queue.new
# a.enqueue 4
# a.enqueue 5
# p a.dequeue
# p a.dequeue
# p a.dequeue
# a.enqueue 4
# p a.dequeue

File: custom_data_structures/js_data_structures/queue.js
var Queue = function(initialValue){
  this.front = null;
  this.back = null;
  if (initialValue){
    var newNode = new Node(initialValue);
    this.front = newNode;
    this.back = newNode;
  }
}

Queue.prototype.enqueue = function(item){
  var newNode = new Node(item);
  if (this.front){
    this.back.next = newNode;
    this.back = newNode;
  } else {
    this.front = newNode;
    this.back = newNode;
  }
}

Queue.prototype.dequeue = function(){
  if (!this.front){
    return null;
  }
  var returnedValue = this.front.value;
  if (this.front.next){
    this.front =this.front.next;
  } else {
    //  queue is empty now
    this.front = null;
    this.last = null;
  }
  return returnedValue;
}

Queue.prototype.peek = function(){
  if (this.front){
    return this.front.value;
  } else {
    return null;
  }
}